### ISSUE TITLE
Reason for smoothing term in numerator of dice loss and score?

### ISSUE THREAD

**Issue by LeonhardFeiner** [qubvel/segmentation_models.pytorch#575]
Why is there a smoothing term in the numerator?
https://github.com/qubvel/segmentation_models.pytorch/blob/8bf52c7e862af006e76a23aae6aa17977d7f9a79/segmentation_models_pytorch/losses/_functional.py#L186

It if the target is zero for all values, the intersection will also be zero.
 
In the metrics case, if output is also fully zero, the output results as _1 = smooth/smooth._
This "helps" to increase overall dice, but in reality the dice for this image/batch should be undefined and not encountered in the average over the dataset. 

For the loss case, it could lead to vanishing predictions for rare classes.
If many batches have output fully zero for one class, the network will get a very optimal result for making the prediction also very close to 0 for every batch. In the case of no target, the loss of the batch is _loss = 1 - smooth/smooth = 0_.
If the network predicts _something  >> smooth > 0_ the loss is _loss = 1 - smooth/something =~ 1_
The jump between 0 and 1 for batches with no target of a rare class is very steep, whereas the slope for batches without class can be smaller.
Therefore, there can be a much higher incentive to get all zero predictions for difficult rare classes than trying to predict this class.

If the smoothing term in the numerator would not be there, the loss for this class would be high, but the network does not have an incentive to make this class likelihood very close to 0.
